taroscript/send.go

```diff
@@ -0,0 +1,163 @@
+	// into more than two outputs.
+	ErrInvalidCollectibleSplit = errors.New(
+		"fund: invalid collectible split",
+	)
+
+	// ErrInvalidChangeOutputLocation is returned when the change output is
+	// not at the expected location (index 0).
+	ErrInvalidChangeOutputLocation = errors.New(
+		"fund: invalid change output location, should be index 0",
+	)
+
+	// ErrInvalidSplitAmounts is returned when the split amounts don't add
+	// up to the amount of the asset being spent.
+	ErrInvalidSplitAmounts = errors.New(
+		"fund: invalid split amounts, sum doesn't match input",
+	)
+)
+
+// PrepareOutputAssets prepares the assets of the given outputs depending on
+// the amounts set on the transaction. If a split is necessary (non-interactive
+// or partial amount send) it computes a split commitment with the given input
+// and spend information. The input MUST be checked as valid beforehand and the
+// change output is expected to be declared as such (and be at index 0).
+func PrepareOutputAssets(input *taropsbt.VInput,
+	outputs []*taropsbt.VOutput) error {
+
+	// This should be caught way earlier but just to make sure that we never
+	// overflow when converting the input amount to int64 we check this
+	// again.
+	if input.Asset.Amount > math.MaxInt64 {
+		return fmt.Errorf("amount int64 overflow")
+	}
+
+	// A collectible cannot be split into individual pieces. So there can
+	// only be a tombstone and a recipient output.
+	if input.Asset.Type == asset.Collectible && len(outputs) > 2 {
+		return ErrInvalidCollectibleSplit
+	}
+
+	var (
+		residualAmount = input.Asset.Amount
+		splitLocators  = make([]*commitment.SplitLocator, len(outputs))
+	)
+	for idx := range outputs {
+		vOut := outputs[idx]
+
+		// We assume the first output is the change output (or
+		// tombstone if there is no change in a non-interactive send).
+		if idx == 0 {
+			// The change output should always be at index 0.
+			if !vOut.IsChange {
+				return ErrInvalidChangeOutputLocation
+			}
+
+			// A zero-amount change output (tombstone) must spend to
+			// the un-spendable NUMS script key.
+			if vOut.Amount == 0 &&
+				!vOut.ScriptKey.PubKey.IsEqual(asset.NUMSPubKey) {
+
+				return commitment.ErrInvalidScriptKey
+			}
+		} else {
+			if vOut.IsChange {
+				return ErrInvalidChangeOutputLocation
+			}
+		}
+
+		residualAmount -= vOut.Amount
+
+		locator := vOut.SplitLocator(input.Asset.ID())
+		splitLocators[idx] = &locator
+	}
+
+	// We should now have exactly zero value left over after splitting.
+	if residualAmount != 0 {
+		return ErrInvalidSplitAmounts
+	}
+
+	// If we have an interactive full value send, we don't need a tomb stone
+	// at all.
+	inputIDCopy := input.PrevID
+	if interactiveFullValueSend(input, outputs) {
+		// We'll now create a new copy of the old asset, swapping out
+		// the script key. We blank out the tweaked key information as
+		// this is now an external asset.
+		outputs[1].Asset = input.Asset.Copy()
```

Comment:
Why index 1 here? Thought this was the case of a 1-in-1-out sweep so it can just modify the same output?

Reply:
Hmm, right. I just assumed that index 0 would always be the change output on the virtual transaction. And if there is no asset change output, we'd just drop it before signing. But it does make sense to only have one output (index 0) in that case. I guess we should implement an itest for the interactive full value send, as that code path is currently never executed outside of the send unit tests.